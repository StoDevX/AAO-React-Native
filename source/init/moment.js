// @flow
import moment from 'moment-timezone'

// These next values made possible by
// http://stackoverflow.com/questions/42513276

// Tell Moment to round up, so that Moment will always report the
// time until something else in a way that matches the clock time.
// That is, at 5:22pm, Moment will report "8 minutes until 5:30pm"
// until the clock ticks over to 5:23pm.
// If we do Math.floor, we go the other direction: it will report
// "7 minutes until 5:30pm" as soon as 5:22pm starts.
moment.relativeTimeRounding(Math.ceil)

// Tell Moment that a minute is 60 seconds, for reporting purposes.
// It defaults to 45 seconds.
moment.relativeTimeThreshold('s', 60)
