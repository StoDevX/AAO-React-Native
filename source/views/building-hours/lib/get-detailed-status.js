// @flow
import moment from 'moment-timezone'
import flatten from 'lodash/flatten'
import type {BuildingType} from '../types'

import {getDayOfWeek} from './get-day-of-week'
import {isChapelTime, formatChapelTime} from './chapel'
import {isScheduleOpenAtMoment} from './is-schedule-open'
import {formatBuildingTimes} from './format-times'

export function getDetailedBuildingStatus(
	info: BuildingType,
	m: moment,
): Array<{isActive: boolean, label: string | null, status: string}> {
	// Friday: 9:00am – Midnight
	// -- or --
	// Friday Breakfast: 7:00am – 9:45am
	// Friday Lunch: 10:30am – 2:00pm
	// Friday Dinner: 4:30pm – 7:30pm
	// -- alternately --
	// Friday Faculty/Staff Swim: 7:00am – 9:45am
	// Friday Lap Swim: 12:45pm – 2:00pm
	// Friday Open Swim: 7:00am – 2:00pm

	let dayOfWeek = getDayOfWeek(m)

	let schedules = info.schedule || []
	if (!schedules.length) {
		return [{isActive: false, label: null, status: 'Hours unknown'}]
	}

	let results = schedules.map(set => {
		let label = set.title
		if (set.closedForChapelTime && isChapelTime(m)) {
			return [
				{
					isActive: false,
					label,
					status: `Closed for chapel: ${formatChapelTime(m)}`,
				},
			]
		}

		let filteredSchedules = set.hours.filter(sched =>
			sched.days.includes(dayOfWeek),
		)
		if (!filteredSchedules.length) {
			return [{isActive: false, label, status: 'Closed today'}]
		}

		return filteredSchedules.map(schedule => {
			let isActive = isScheduleOpenAtMoment(schedule, m)
			let status = formatBuildingTimes(schedule, m)
			if (set.isPhysicallyOpen === false) {
				isActive = false
			}
			return {isActive, label, status}
		})
	})

	return flatten(results)
}
