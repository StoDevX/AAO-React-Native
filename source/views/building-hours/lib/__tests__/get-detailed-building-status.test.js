// @flow
import {getDetailedBuildingStatus} from '../get-detailed-status'
import {plainMoment} from './moment.helper'

it('returns a list of [isOpen, scheduleName, verboseStatus] tuples', () => {
	let m = plainMoment('06-23-2018 1:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [
					{days: ['Mo', 'Tu', 'We', 'Th'], from: '10:30am', to: '12:00am'},
					{days: ['Fr', 'Sa'], from: '10:30am', to: '2:00am'},
					{days: ['Su'], from: '10:30am', to: '12:00am'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)

	expect(actual).toBeInstanceOf(Array)
	expect(actual.length).toEqual(1)
	expect(typeof actual[0].isActive).toBe('boolean')
	expect(typeof actual[0].label).toBe('string')
	expect(typeof actual[0].status).toBe('string')

	expect(actual).toMatchSnapshot()
})

it('checks a list of schedules to see if any are open', () => {
	let m = plainMoment('06-23-2018 1:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [
					{days: ['Mo', 'Tu', 'We', 'Th'], from: '10:30am', to: '12:00am'},
					{days: ['Fr', 'Sa'], from: '10:30am', to: '2:00am'},
					{days: ['Su'], from: '10:30am', to: '12:00am'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)
	expect(actual).toMatchSnapshot()

	expect(actual[0].isActive).toBe(true)
})

it('handles multiple internal schedules for the same timeframe', () => {
	let m = plainMoment('06-18-2018 1:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [
					{days: ['Mo'], from: '10:30am', to: '12:00pm'},
					{days: ['Mo'], from: '1:00pm', to: '3:00pm'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)
	expect(actual).toMatchSnapshot()

	expect(actual[0].isActive).toBe(false)
	expect(actual[1].isActive).toBe(true)
})

it('handles multiple named schedules for the same timeframe', () => {
	let m = plainMoment('06-18-2018 1:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [{days: ['Mo'], from: '10:30am', to: '12:00pm'}],
			},
			{
				title: 'Hours2',
				hours: [
					{days: ['Mo'], from: '10:30am', to: '12:00pm'},
					{days: ['Mo'], from: '1:00pm', to: '3:00pm'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)
	expect(actual).toMatchSnapshot()

	expect(actual[0].isActive).toBe(false)
	expect(actual[1].isActive).toBe(false)
	expect(actual[2].isActive).toBe(true)
})

it('returns false if none are available for this day', () => {
	let m = plainMoment('06-17-2018 1:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [
					{days: ['Mo', 'Tu', 'We', 'Th'], from: '10:30am', to: '12:00am'},
					{days: ['Fr', 'Sa'], from: '10:30am', to: '2:00am'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)
	expect(actual).toMatchSnapshot()

	expect(actual[0].isActive).toBe(false)
})

it('returns false if none are open', () => {
	let m = plainMoment('06-19-2018 3:00pm', 'MM-DD-YYYY h:mma')
	let building = {
		name: 'building',
		category: '???',
		breakSchedule: {},
		schedule: [
			{
				title: 'Hours',
				hours: [
					{days: ['Mo', 'Tu', 'We', 'Th'], from: '10:30am', to: '2:00pm'},
					{days: ['Fr', 'Sa'], from: '10:30am', to: '2:00pm'},
				],
			},
		],
	}

	const actual = getDetailedBuildingStatus(building, m)
	expect(actual).toMatchSnapshot()

	expect(actual[0].isActive).toBe(false)
})
