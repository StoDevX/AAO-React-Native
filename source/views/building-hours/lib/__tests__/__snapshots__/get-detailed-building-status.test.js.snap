// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks a list of schedules to see if any are open 1`] = `
Array [
  Object {
    "isActive": true,
    "label": "Hours",
    "status": "10:30 AM — 2:00 AM",
  },
]
`;

exports[`handles multiple internal schedules for the same timeframe 1`] = `
Array [
  Object {
    "isActive": false,
    "label": "Hours",
    "status": "10:30 AM — Noon",
  },
  Object {
    "isActive": true,
    "label": "Hours",
    "status": "1:00 PM — 3:00 PM",
  },
]
`;

exports[`handles multiple named schedules for the same timeframe 1`] = `
Array [
  Object {
    "isActive": false,
    "label": "Hours",
    "status": "10:30 AM — Noon",
  },
  Object {
    "isActive": false,
    "label": "Hours2",
    "status": "10:30 AM — Noon",
  },
  Object {
    "isActive": true,
    "label": "Hours2",
    "status": "1:00 PM — 3:00 PM",
  },
]
`;

exports[`returns a list of [isOpen, scheduleName, verboseStatus] tuples 1`] = `
Array [
  Object {
    "isActive": true,
    "label": "Hours",
    "status": "10:30 AM — 2:00 AM",
  },
]
`;

exports[`returns false if none are available for this day 1`] = `
Array [
  Object {
    "isActive": false,
    "label": "Hours",
    "status": "Closed today",
  },
]
`;

exports[`returns false if none are open 1`] = `
Array [
  Object {
    "isActive": false,
    "label": "Hours",
    "status": "10:30 AM — 2:00 PM",
  },
]
`;
